scalding-core/src/main/scala/com/twitter/scalding/ExecutionOptimizationRules.scala

```diff
@@ -0,0 +1,157 @@
+          case (e: Execution.FlatMapped[a, b], f) =>
+            Literal.Unary(f(e.prev), Execution.FlatMapped(_: Execution[a], e.fn))
+          case (e: Execution.Zipped[a, b], f) =>
+            Literal.Binary(f(e.one), f(e.two), Execution.Zipped(_: Execution[a], _: Execution[b]))
+        }
+      }
+    )
+
+  object ZipWrite extends PartialRule[Execution] {
+    type InputType = (Config, Mode, Execution.Writer, ConcurrentExecutionContext)
+    type ResFun[T] =  InputType => Future[T]
+
+    case class NewFn[T, U](left: ResFun[T], right: ResFun[U]) extends (InputType => Future[(T, U)]) {
+      override def apply(tup: InputType): Future[(T, U)] =
+        Execution.failFastZip(left(tup), right(tup))(tup._4)
+    }
+
+    override def applyWhere[T](on: Dag[Execution]) = {
+      /*
+      * run this and that in parallel, without any dependency. This will
+      * be done in a single cascading flow if possible.
+      *
+      * If both sides are write executions then merge them
+      */
+      case Execution.Zipped(Execution.WriteExecution(oH, oT, oR), Execution.WriteExecution(tH, tT, tR)) =>
+        Execution.WriteExecution(oH, tH :: tT ::: oT, NewFn(oR, tR))
+    }
+  }
+
+  object ZipMap extends PartialRule[Execution] {
+    case class MapLeft[S, T, B](fn: S => B) extends (((S, T)) => (B, T)) {
+      override def apply(st: (S, T)): (B, T) = (fn(st._1), st._2)
+    }
+
+    case class MapRight[S, T, B](fn: T => B) extends (((S, T)) => (S, B)) {
+      override def apply(st: (S, T)): (S, B) = (st._1, fn(st._2))
+    }
+
+    override def applyWhere[T](on: Dag[Execution]) = {
+      case Execution.Zipped(Execution.Mapped(left, fn), right) =>
+        Execution.Zipped(left, right).map(MapLeft(fn))
+      case Execution.Zipped(left, Execution.Mapped(right, fn)) =>
+        Execution.Zipped(left, right).map(MapRight(fn))
+    }
+  }
+
+  object ZipFlatMap extends PartialRule[Execution] {
+    case class LeftZipRight[S, T, B](left: Execution[B], fn: S => Execution[T]) extends (S => Execution[(B, T)]) {
+      private val fun = fn.andThen(left.zip)
+
+      override def apply(s: S): Execution[(B, T)] = fun(s)
+    }
+
+    case class RightZipLeft[S, T, B](right: Execution[B], fn: S => Execution[T]) extends (S => Execution[(T, B)]) {
+      private val fun = fn.andThen(_.zip(right))
+
+      override def apply(s: S): Execution[(T, B)] = fun(s)
+    }
+
+    case class NestedZip[S, T, B, A](right: Execution[B], lfn: S => Execution[T], rfn: B => Execution[A]) extends (S => Execution[(T, A)]) {
+      private val fun = lfn.andThen { lr =>
+        Execution.FlatMapped(right, rfn.andThen(lr.zip))
+      }
+
+      override def apply(s: S): Execution[(T, A)] = fun(s)
+    }
+
+    override def applyWhere[T](on: Dag[Execution]) = {
+      case Execution.Zipped(Execution.FlatMapped(left, lfn), Execution.FlatMapped(right, rfn)) =>
+        Execution.FlatMapped(left, NestedZip(right, lfn, rfn))
```

Comment:
I think we are breaking a lot of parallelism here. I'm worried this will be a huge regression for us since this now runs in series.

I think we should probably only do this transformation if `left` is a "fast" Execution (say, fromFuture, getArgs, etc...), which we assume happens fast and we should see if it fails before we start any map reduce job.

Previously, left and right would be started at the same time here.

We use Executions pretty heavily and I don't think we have good test coverage on parallelism regressions like this.

Reply:
maybe we can write a function like: `def isFastExecution[A](ex: Execution[A])` which is something like:
```scala
case FutureConst(_) => true
case GetCounters(e) => isFastExecution(e)
case Mapped(e) => isFastExecution(e)
case WriteExecution(_, _, _, _) => false
case Zipped(a, b) => isFastExecution(a) && isFastExecution(b)
...
```
This can be made stack safe by using a list:
```scala
@annotation.tailrec
def areFastEx(es: List[Execution[Any]]): Boolean = 
  es match {
    case None => true
    case h :: tail =>
      h match {
        case WriteExecution(_, _, _, _) => false
        case Zipped(a, b) => areFastEx(a :: b :: tail)
        case FutureConst(_) => areFastEx(tail)
        ...
      }
  }
```

Reply:
We probably have to say a `FlatMapped` is not fast, since we have no idea what the inner execution is, but I think that is still okay, lots of examples we have will just be zips, writes, maps, and fromFuture/getArgs....

So, I guess the law we want is we never serialize two Writes that were previously parallel, but otherwise we don't mind serializing FutureConst (since we assume that is O(1-10000ms) whereas a WriteExecution is O(10m - 10h).

Reply:
good, point. thanks to catching this.